### ISSUE TITLE
MVP new Actor Framework & Scheduler

### ISSUE THREAD

**Issue by meyerdan** [zeebe-io/zeebe#635]
# Zeebe's Workload

Zeebe has a diverse work load including

* processing client requests
* serving subscribers in streaming mode
* running background jobs (compaction, ...)

Also, it needs to participate in network protocols such as Gossip and Raft which formulate requirements like sending pings and heartbeats at regular intervals or responding to such requests within a particular time frame.

# Requirements

This results in conflicting requirements as to how to process this workload

* **Maximize utilization**: Utilize system resources as much as possible up to saturation point
* **Meet deadlines**: Ensure that even under high load, tasks with fixed deadlines (such as Gossip's ping requests or Raft's append entries requests are sent withing the required deadlines)
* **Ensure sustainable operation**: even under high request load, subscriptions and background jobs still need to get a fair share of the available resources.

# Actors and Green Threads

Currently Zeebe uses an actor model based on green threads. The system is implemented as number number of cooperating, non-blocking actors which are scheduled to a small number of threads. To not oversubscribe the CPUs (utilize past saturation point), the number of threads needs to be smaller or equal to the number of cores the system has. This is highly efficient since the actors are non-blocking: when a actor cannot make progress because it is waiting on data to become available or other conditions, it released the thread which can now progress to do work inside another actor.

While this model is generally very well suited for implementing a high-throughput system like Zeebe, the current implementation has many problems.

# Problems with the current Implementation 

The current implementation has the following problems

* Poor load balancing: when running on systems with a larger number of cores (>=4) the system makes a poor job at balancing load across these cores to maximize utilization
* Polling is the only solution to waiting: whenever an actor is waiting on some condition, it needs to poll for the condition to become true
* No possibility to run blocking tasks/polling
* Heavy on boilerplate for recurring tasks like implementing timeouts, waiting on the completion of futures, ... a lot of boilerplate code needs to be written all the time
* Only coarsely grained control: from the outside it is only possible to call the actor's doWork() method, but we have no control over the inner workings of the actor. This makes it impossible to implement I/O schedulers and the like.

# Proposed Implementation

The proposed implementation is comprised of two parts:

* More expressive programming model: considerably reduces boilerplate and provides fine grained control to the scheduler
* Efficient, O(1) workstealing scheduler supporting the more expressive programming model

In more detail, this means

* Programming model allowing to directly express
    * consuming from "byte sources"
    * running periodic tasks
    * dealing with backpressure (running tasks until they are done, yielding CPU/thread time)
    * runing tasks with a fixed delay
    * blocking on Features to complete
    * running blocking tasks / polling
    * ...
* Load-balancing through work-stealing (based on wait-free queue algorithms)
* Built-in facility for timers (based on timer wheel facility)
* Task priority classes with quotas to ensure that the system focuses on high priority tasks while ensuring that lower priority tasks are executed up to an assignable quota
* Metrics allowing to track and eliminate "jumbo tasks" taking up too much runtime and blocking other tasks and by this preventing the system to meet it's deadlines.
* Possibility to run blocking tasks and polling in separate threads blocking in the OS

# Tradeoffs made

We can never have everyting at the same time. The following tradeoffs are made in the implementation 

* Garbage Freedom: the implementation is garbage free only in constructs which we expect to be used in the datapath 
* The scheduler expects a smaller amount of actors (hundreds) running a long time spawning many jobs (millions). Scheduling a new actor can be slower while running jobs needs to be super fast

# TODOS for MVP

- [x] onActorStarted(...)
- [x] run(...), runDelayed(...), runAtfixedRate(...)
- [x] basic metrics
- [x] handle task runner idle
- [x] callables, futures and await
- [x] consume(...)
- [x] poll blocking
- [x] runBlocking(...) - short-lived blocking task
- [x] condition variables
- [x] runUntilDone
- [x] onActorClose(...)
- [x] ensure subscriptions do not interrupt current job 

# TODOs for full version

- [x] task execution metrics #671
- [x] metrics for detecting "jumbo" tasks #673
- [x] priority classes and priority quotas #674
- [x] support for I/O-bound actors #679 
- [ ] formalize lifecycle (open should not be interruptible by close)
- [ ] task-quotas and preemption of long-running tasks 
- [ ] possibility to await regular futures (via blocking)
- [ ] testcases
- [ ] exception & error handling (particularly, look into agrona's DistinctErrorLog log)
- [ ] benchmarks
- [ ] improve code (decide on package visibility and direct field access, document most important concepts)
- [ ] optimizations (branches!, check false sharing, evaluate thread affinity) 
- [ ] reduce garbage
- [ ] logging
- [ ] logging: custom MDC properties (see #574)

**Comment by meyerdan** [comment]
@menski, I think it makes sense to merge scheduler mvp to master, it is not used yet and it does not change anything else, hence it should not cause any  problems.

In order to do that, we need to update agrona to 0.9.12. The reason is that we are using DeadlineTimerWheel which has been added in that version.
